src/Microsoft.DotNet.Interactive.HttpRequest/HttpRequestKernel.cs

```diff
@@ -269,192 +234,61 @@ private static HttpRequestMessage GetRequestMessage(ParsedHttpRequest parsedRequ
-                currentDiagnostics = new List<Diagnostic>();
-            }
-
-            var rebuiltLine = new StringBuilder();
-            var lastStart = 0;
-            var interpolationStart = lineText.IndexOf(InterpolationStartMarker);
-            while (interpolationStart >= 0)
-            {
-                rebuiltLine.Append(lineText[lastStart..interpolationStart]);
-                var interpolationEnd = lineText.IndexOf(InterpolationEndMarker, interpolationStart + InterpolationStartMarker.Length);
-                if (interpolationEnd < 0)
-                {
-                    // no end marker
-                    // TODO: error?
-                }
-
-                var variableName = lineText[(interpolationStart + InterpolationStartMarker.Length)..interpolationEnd];
-                if (_variables.TryGetValue(variableName, out var value))
-                {
-                    rebuiltLine.Append(value);
-                }
-                else
-                {
-                    // no variable found; keep old code and report diagnostic
-                    rebuiltLine.Append(InterpolationStartMarker);
-                    rebuiltLine.Append(variableName);
-                    rebuiltLine.Append(InterpolationEndMarker);
-                    var position = new LinePositionSpan(
-                        new LinePosition(line, interpolationStart + InterpolationStartMarker.Length),
-                        new LinePosition(line, interpolationEnd));
-                    currentDiagnostics.Add(new Diagnostic(position, DiagnosticSeverity.Error, "HTTP404", $"Cannot resolve symbol '{variableName}'"));
-                }
+        var diagnostics = new List<Diagnostic>();
 
-                lastStart = interpolationEnd + InterpolationEndMarker.Length;
-                interpolationStart = lineText.IndexOf(InterpolationStartMarker, lastStart);
-            }
-
-            rebuiltLine.Append(lineText[lastStart..]);
-            currentLines.Add(rebuiltLine.ToString());
-        }
-
-        if (MightContainRequest(currentLines))
+        foreach (var diagnostic in parseResult.GetDiagnostics())
         {
-            var requestCode = string.Join('\n', currentLines);
-            result.Add((requestCode, currentDiagnostics));
+            diagnostics.Add(diagnostic);
         }
 
-        return result;
-    }
-
-    private static bool MightContainRequest(IEnumerable<string> lines)
-    {
-        return lines.Any(line => IsRequest.IsMatch(line));
-        //return lines.Any() && lines.Any(line => !string.IsNullOrWhiteSpace(line));
-    }
-
-    private IEnumerable<ParsedHttpRequest> ParseRequests(string requests)
-    {
-        var parsedRequests = new List<ParsedHttpRequest>();
-
-        foreach (var (request, diagnostics) in InterpolateAndGetDiagnostics(requests))
+        foreach (var expressionNode in parseResult.SyntaxTree.RootNode.DescendantNodesAndTokensAndSelf().OfType<HttpExpressionNode>())

```

Comment:
> Some of those diagnostics may be reported in URIs / headers with no embedded expressions (but would still be diagnostics that are reported at bind-time).

The block before this (`parseResult.GetDiagnostics()`) is getting the syntactic diagnostics and this code is getting the symbolic diagnostics.  My thinking was that keeping them as separate steps is useful in this code path (i.e. in the context of `RequestDiagnostics`) because the latter will sometimes include side effects such as UI prompts which we shouldn't trigger here.

Reply:
What I meant is that there could be more to the symbolic diagnostics beyond variable resolution and this loop only seems to be handling variable resolution.

The `TryGetRequest` API (and other APIs it calls such as `TryGetUri`) would already be considering all such diagnostics - so my thinking was that it may be better to rely exclusively on that API and centralize all validation / binding logic over there.

Reply:
> What I meant is that there could be more to the symbolic diagnostics beyond variable resolution and this loop only seems to be handling variable resolution.

Do we have an example of this?

Reply:
One example may be if we add validation for supported version numbers for `HttpVersionNode`. Another may be if we need to add validation for specific values that are supported / unsupported for `HeaderValueNode`.

Wonder if we should try and create a more general `TryBind()` API which performs the same binding steps that `TryGetHttpRequest()` does without creating / returning the `System.Net.Http.*` objects. However, I feel some of these validations may require calling into `System.Net.Http` regardless so we may not be saving much...

Reply:
I would expect these to be directly handled by the parser, since they don't change often and they don't vary based on the host environment.